app/src/main/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddleware.kt

```diff
@@ -40,6 +40,7 @@ class MessagingMiddleware(
                 coroutineScope.launch {
                     val messages = messagingStorage.getMessages()
                     context.store.dispatch(UpdateMessages(messages))
+                    context.store.dispatch(Evaluate)
```

Comment:
I'm a bit confused if this will reliably sole the issue,  as the `Evaluate` action [will trigger an UI update](https://github.com/mozilla-mobile/fenix/blob/49ccb63dd2232464e9e1a3d96d4f47f670b25c7e/app/src/main/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddleware.kt#L46-L54), but here we are initializing the app when we haven't loaded the  `HomeFragment` where we are showing the message.

Reply:
The race was between 
- [this callback from `FenixApplication`](https://github.com/mozilla-mobile/fenix/blob/49ccb63dd2232464e9e1a3d96d4f47f670b25c7e/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#L456) which [would dispatch MessagingAction.Restore](https://github.com/mozilla-mobile/fenix/blob/49ccb63dd2232464e9e1a3d96d4f47f670b25c7e/app/src/main/java/org/mozilla/fenix/FenixApplication.kt#L472) and 
- `MessagingFeature` which would dispatch [MessagingAction.Evaluate](https://github.com/mozilla-mobile/fenix/blob/49ccb63dd2232464e9e1a3d96d4f47f670b25c7e/app/src/main/java/org/mozilla/fenix/gleanplumb/MessagingFeature.kt#L19)

With some logs I saw yesterday together with James that `MessagingAction.Evaluate` would be dispatched first by `MessagingFeature` before the Nimbus callback would dispatch `MessagingAction.Restore`.
And as such we would try to show messages before they are available.

The changes from this patch will help ensure that `MessagingAction.Evaluate` is always called after `MessagingAction.Restore` and `MessagingAction.Update`, after the messages are available.

Reply:
This changes will help with the scenario in which `HomeFragment` was loaded but `MessagingAction.Restore` would not yet be dispatched.
With this patch we are dispatching `MessagingAction.Evaluate` a second time to complete the first one which didn't show anything because messages were not yet available when `HomeFragment` is shown.

Reply:
I'm less familiar with the unidirectional data flow, and where the Middleware sits in the stack, so am happy to defer to @Amejia481 's good judgement here.

Reply:
The only drawback is as we are dispatching twice the `MessagingAction.Restore`, in some cases we would be reduce the times that users will see a message, ideally we would improve that by observing the `messages` in `MessagingFeature`, and dispatching when they are updated.

Reply:
For now, I think this address the main issue, we could do a follow up ticket to the solution improve it!

Reply:
This patch is only needed for 108.
109 (current Nightly soon to become Beta) and forward switched from depending on that Nimbus callback and so won't need any change.

Reply:
Thank you!
Created https://github.com/mozilla-mobile/fenix/issues/28156 to evaluate and fix in later versions also.

Reply:
Added a comment in code to offer a bit more context for why `Evaluate` has to be called after `Restore`.

Reply:
Perfect, feel free to land when you are ready!